src/sources/logstash.rs

```diff
@@ -0,0 +1,762 @@
+use flate2::read::ZlibDecoder;
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::{
+    collections::{BTreeMap, VecDeque},
+    convert::TryFrom,
+    io::{self, Read},
+};
+use tokio_util::codec::Decoder;
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct LogstashConfig {
+    address: SocketListenAddr,
+    keepalive: Option<TcpKeepaliveConfig>,
+    tls: Option<TlsConfig>,
+    receive_buffer_bytes: Option<usize>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<LogstashConfig>("logstash")
+}
+
+impl GenerateConfig for LogstashConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: SocketListenAddr::SocketAddr("0.0.0.0:5044".parse().unwrap()),
+            keepalive: None,
+            tls: None,
+            receive_buffer_bytes: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "logstash")]
+impl SourceConfig for LogstashConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let source = LogstashSource {};
+        let shutdown_secs = 30;
+        let tls = MaybeTlsSettings::from_config(&self.tls, true)?;
+        source.run(
+            self.address,
+            self.keepalive,
+            shutdown_secs,
+            tls,
+            self.receive_buffer_bytes,
+            cx.shutdown,
+            cx.out,
+        )
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "logstash"
+    }
+
+    fn resources(&self) -> Vec<Resource> {
+        vec![self.address.into()]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LogstashSource {}
+
+impl TcpSource for LogstashSource {
+    type Error = DecodeError;
+    type Decoder = LogstashDecoder;
+
+    fn decoder(&self) -> Self::Decoder {
+        LogstashDecoder::new()
+    }
+
+    // https://github.com/logstash-plugins/logstash-input-beats/blob/master/PROTOCOL.md#ack-frame-type
+    fn build_ack(&self, frame: &LogstashEventFrame) -> Bytes {
+        let mut bytes: Vec<u8> = Vec::with_capacity(6);
+        bytes.push(frame.protocol.into());
+        bytes.push(LogstashFrameType::Ack.into());
+        bytes.extend(frame.sequence_number.to_be_bytes().iter());
+        Bytes::from(bytes)
+    }
+
+    fn build_event(&self, frame: LogstashEventFrame, host: Bytes) -> Option<Event> {
+        let mut log = LogEvent::from(
+            frame
+                .fields
+                .into_iter()
+                .map(|(key, value)| (key, Value::from(value)))
+                .collect::<BTreeMap<_, _>>(),
+        );
+        if log.get(log_schema().host_key()).is_none() {
+            log.insert(log_schema().host_key(), host);
+        }
+        if log.get(log_schema().timestamp_key()).is_none() {
+            if let Some(timestamp) = log.get("@timestamp") {
+                let timestamp = timestamp.clone();
```

Comment:
Is this shadowing the `timestamp` on the line above?

Reply:
It is. Variable shadowing is a common practice in Rust (when it makes sense to do so).